Data module

pheuer opened the pull request:
This PR creates a new module in `plasmapy/utils` for downloading files from the [PlasmaPy data repository](https://github.com/PlasmaPy/data). I see two main use-cases for this code: 

- Accessing data files for use in example notebooks and CI tests (as in #1274 )
- Downloading file required for future functionality, e.g. lookup tables of atomic data, maybe. 

Originally I tried using [the `parfive` package](https://parfive.readthedocs.io/en/stable/) to do this, but I ran into a number of issues. Files downloaded from GitHub have a "Content-Encoding:'gzip'" header, but are not actually compressed, causing parfive to raise an exception. Since this is such a simple downloader, I was able to just avoid the problem by using the `requests` module to handle the download directly and ignore the misleading header. 

This PR is required for #1274 : originally I was working on this module there but I've split it off here for easier reviewing.

namurphy reviewed (changes requested):
Thank you for doing this!  I made a few comments and suggestions.  It will be incredibly helpful to have a place where we can put sample data, so I very much appreciate this.  I expect this will be particularly helpful for example notebooks. 

I have a couple of broader thoughts, though these are out-of-scope for this PR.
 - I'm wondering about having a naming scheme that is consistent with SunPy, since we have a lot of potential users from the heliophysics community.  They have `sunpy.data` for sample data used in examples (including some downloader stuff, I think), `sunpy.net` for downloading from repositories, and `sunpy.io` for file input/output.  Adding a new subpackage would mean we'd have to amend PLEP 7, so all of this is out-of-scope for this PR.
 - If the files are small enough, I'd have no problem with including them in PlasmaPy directly.  That would probably make it easier to use that data in notebooks.
 - In the longer term, I'm wondering if it would be best to have a place to store this data in addition to GitHub, since links on GitHub are not guaranteed to be persistent.  SunPy has https://data.sunpy.org, and another possibility would be to archive them on Zenodo...though there could be some unforeseen problems there.  Again, this is waaaay out of scope for this PR, so it would probably make sense to create a separate issue regarding this.

Thank you again!

pheuer commented on plasmapy/utils/data/data.py:
```diff
@@ -0,0 +1,76 @@
+        #
+        # Missing files on GitHub will resolve to a 404 html page, so we use
+        # this as an indicator that the file may not exist.
+        allowed_types = ["text/plain; charset=utf-8", "image/png"]
```
Absolutely - we'd just have to see what Content-Type GitHub assigns them and make sure they are passed through.

pheuer replied:
Actually, I think the better option is just to allow everything through except html files, which we will interpret as the file not being found (eg. GitHub returned a 404 page). I've changed to that and tested that it works with an h5 file.

pheuer commented on plasmapy/utils/data/data.py:
```diff
@@ -0,0 +1,76 @@
+"""
+Contains functionality for downloading files from a URL. Intended for
+downloading files from the PlasmaPy data repository.
+
+"""
```
Good idea - done

pheuer commented on plasmapy/utils/data/tests/test_data.py:
```diff
@@ -0,0 +1,46 @@
+            assert np.allclose(arr[0, :], np.array([1e-3, 1.043e2]))
+
+        # Get the file again, already existing so it doesn't download it again
+        path = data.get_file(filename)
```
Yep, done

pheuer commented on plasmapy/utils/data/data.py:
```diff
@@ -0,0 +1,76 @@
+        url = urljoin(base_url, basename)
+
+        # Get the requested content
+        r = requests.get(url)
```
Good point - changing this to "reply"

pheuer commented on plasmapy/utils/data/tests/test_data.py:
```diff
@@ -0,0 +1,46 @@
+
+    """
+
+    # Delete file if it already exists, so the test always downloads it
```
Ok yep, now it uses `tmp_dir` which is defined by pytest.

pheuer commented:
> Thank you for doing this! I made a few comments and suggestions. It will be incredibly helpful to have a place where we can put sample data, so I very much appreciate this. I expect this will be particularly helpful for example notebooks.

Thanks for the comments!

> I have a couple of broader thoughts, though these are out-of-scope for this PR.
> 
> * I'm wondering about having a naming scheme that is consistent with SunPy, since we have a lot of potential users from the heliophysics community.  They have `sunpy.data` for sample data used in examples (including some downloader stuff, I think), `sunpy.net` for downloading from repositories, and `sunpy.io` for file input/output.  Adding a new subpackage would mean we'd have to amend PLEP 7, so all of this is out-of-scope for this PR.

I agree that this ultimately makes sense, especially as more open data becomes available.

> * If the files are small enough, I'd have no problem with including them in PlasmaPy directly.  That would probably make it easier to use that data in notebooks.

Noted - I'd be more tempted to do that if I intended users to ever use these directly (these are just for tests/examples). Maybe for lookup tables or something incorporating them directly makes sense, so no one is ever on a plane and suddenly realizes they can't use XYZ function.

> * In the longer term, I'm wondering if it would be best to have a place to store this data in addition to GitHub, since links on GitHub are not guaranteed to be persistent.  SunPy has https://data.sunpy.org, and another possibility would be to archive them on Zenodo...though there could be some unforeseen problems there.  Again, this is waaaay out of scope for this PR, so it would probably make sense to create a separate issue regarding this.

That would be great: one thing I've learned working on this is that GitHub definitely isn't the best way to store datafiles. SunPy's downloader checks several mirrors for the given filenames, so we could potentially have several sources to pull from.

pheuer commented on plasmapy/utils/data/data.py:
```diff
@@ -0,0 +1,76 @@
+
+    base_url : str, optional
+        The base URL of the file to be downloaded. Defaults to the main
+        directory of the PlasmaPy data repository.
```
Done

namurphy commented:
> SunPy's downloader checks several mirrors for the given filenames, so we could potentially have several sources to pull from.

Oh good idea!  Having mirrors is something we ought to eventually be doing, most likely.  Though not magnetic mirrors, since those are lossy.

namurphy commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,75 @@
+
+    base_url : str, optional
+        The base URL of the file to be downloaded. Defaults to the main
+        directory of the PlasmaPy data repository.
```
```suggestion
        directory of the `PlasmaPy data repository`_.
```
[optional] I'm wondering if it would be useful to define a reST shortcut in `docs/common_links.rst` so that we can easily have hyperlinks for it in the online documentation.

pheuer replied:
Done - not sure if I did it right?

namurphy commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,75 @@
+        directory of the PlasmaPy data repository.
+
+    directory : str, optional
+        The full path to the desired download location. Defaults to the
+        default PlasmaPy data download directory.
```
[optional] It might be helpful to state the location of the data download directory here.

pheuer replied:
Done - let me know if there's a better way to format that.

pheuer commented:
Right - no leaky mirrors here! 

I tried to add the new module to the docs, but I don't think I'm doing that right...where do doc files for utils go?

namurphy reviewed (commented):
It looks like there still needs to be a stub file for the module in `docs/api_static`. We recently added a discussion about that at the bottom of [this section in our doc guide](https://docs.plasmapy.org/en/latest/development/doc_guide.html#narrative-documentation-guidelines).

namurphy commented:
What would you think of changing the name of the data repository from `data` to something like `PlasmaPy-data`?  My motivation is that if someone does an `ls` in a directory containing a whole bunch of repositories, it wouldn't immediately be clear what `data` this referred to.  SunPy calls theirs `sample-data`

pheuer commented:
> What would you think of changing the name of the data repository from `data` to something like `PlasmaPy-data`? My motivation is that if someone does an `ls` in a directory containing a whole bunch of repositories, it wouldn't immediately be clear what `data` this referred to. SunPy calls theirs `sample-data`

Good idea: I've changed it to `PlasmaPy-data`

I've added stub files for utils/data and utils/data/downloader: not sure if I need both. I did add a automodapi reference in the index for the utils module, so does that count as running on utils/data and I just need a stub file for the downloader module?

rocco8773 commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,76 @@
+    if "." not in str(basename):
+        raise ValueError(f"'filename' ({basename}) must include an extension.")
+
+    if directory is None:
```
```suggestion
    if directory is None:  # coverage: ignore
```

StanczakDominik commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,76 @@
+
+    path = os.path.join(directory, basename)
+
+    # If file doesn't exist locally, download it
```
One limitation of this that I'm sure has already been solved elsewhere is that if the upstream file changes, but does not change the filename, this will not update the local version. Could be good to note that in the docstring!

One way of solving that in the future would be to note the SHA1 checksum in the filename, e.g. store "NIST.dat" as "NIST-1234567890.dat" online, and if the file exists... Yeah, this is getting a little too complex for this train ride; it'd be good to investigate how sunpy or astropy or sklearn solve this. Future work, though.

pheuer replied:
I've been playing around with this on [another branch](https://github.com/PlasmaPy/PlasmaPy-data/tree/data_with_hashes), but I think we should leave it for a future PR. For now this module is exclusively for notebooks + testing, so we can just manage it manually for now.

StanczakDominik reviewed (approved):
Looks great for what it's meant to do!

pheuer commented:
@namurphy This is done (fixed the final docs errors): can we get it merged in? 

I explored @StanczakDominik 's idea of using file hashes to check if files on the server had been updated, but I haven't decided exactly how to do that yet. Since this is just for tests + example notebook files right now, I think we should leave that for a future PR (I have a branch set up to keep playing with the idea).

namurphy reviewed (changes requested):
Thank you for working on this, @pheuer! Most of my suggestions are either pretty minor, or for things that I'm thinking would be best to follow up on in a future issue & PR instead of this one. This is going to be useful functionality to have!

pheuer commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,76 @@
+# by the get_file function.
+_BASE_URL = "https://github.com/PlasmaPy/PlasmaPy-data/raw/main/data/"
+
+# TODO: use a config file variable to allow users to set a location for this folder?
```
Yes, absolutely!

pheuer commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,75 @@
+    directory : str, optional
+        The full path to the desired download location. Defaults to the
+        default PlasmaPy data download directory
+        plasmapy/utils/data/downloads/
```
Wow there really is a sphinx syntax for everything!

pheuer commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,75 @@
+    directory : str, optional
+        The full path to the desired download location. Defaults to the
+        default PlasmaPy data download directory
+        plasmapy/utils/data/downloads/
```
Not sure I understand the issue with root vs non-root, but this does sound nicer than keeping that data in the module, so I've made this change. 

Note that it isn't tested because the tests save data to a temporary directory...

pheuer commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,75 @@
+    if directory is None:  # coverage: ignore
+        directory = _DOWNLOADS_PATH
+
+    path = os.path.join(directory, basename)
```
This was very easy so I just made the change now.

pheuer commented on setup.cfg:
```diff
@@ -56,6 +56,7 @@ install_requires =
     tqdm >= 4.41.0
     voila >= 0.2.15
     xarray >= 0.15.0
+    requests >= 2.27.1
```
Yep, done

namurphy reviewed (approved):
LGTM!  And thank you for doing this!

namurphy commented on plasmapy/utils/data/downloader.py:
```diff
@@ -0,0 +1,78 @@
+        raise ValueError(f"'filename' ({basename}) must include an extension.")
+
+    if directory is None:  # coverage: ignore
+        directory = Path(Path.home(), ".plasmapy", "downloads")
```
I was wondering how getting the home directory would be handled with different operating systems, but it looks like it's something that pathlib takes care of for us!

pheuer replied:
Yep!

pheuer commented:
Thanks @namurphy !
